File: Week3/Assign3.1P/CounterTaskcontinued/CounterTask/CounterTask/Counter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CounterTask
{
    public class Counter
    {
        private int _count;
        private string _name;

        public string Name
        {
            get
            {
                return _name;
            }
            set
            {
                _name = value;
            }
        }

        public int Ticks
        {
            get
            {
                return _count;
            }
        }


        public Counter(string name)
        {
            _name = name;
            _count = 0;
        }

        public void Increment()
        {
            try
            {
                checked
                {
                    _count++;
                }
            }
            catch (OverflowException e)
            {
                Console.WriteLine("Overflow detected: " + e.Message);
            }
        }
        public void Reset()
        {
            _count = 0;
        }

        public void ResetByDefault()
        {
            unchecked
            {
                _count = (int)2147483647447;
            }
        }


    }
}

File: Week11/11.1/Counter.js
class Counter {
    constructor(name) {
        this._name = name;
        this._count = 0;
    }

    get name() {
        return this._name;
    }

    set name(value) {
        this._name = value;
    }

    get ticks() {
        return this._count;
    }

    increment() {
        try {
            this._count++;
        } catch (e) {
            console.error("Overflow detected: " + e.message);
        }
    }

    reset() {
        this._count = 0;
    }

    resetByDefault() {
        // JavaScript uses floating point numbers, overflow is not an issue in the same way
        this._count = 2147483647; // Use max 32-bit int value instead
    }
}
